File: 1358. Number of Substrings Containing All Three Characters/main.py
from collections import defaultdict


class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        cnt = defaultdict(int)
        res = 0

        l = 0
        for r in range(len(s)):
            cnt[s[r]] += 1
            while len(cnt) == 3:
                res += len(s) - r
                cnt[s[l]] -= 1
                if cnt[s[l]] == 0:
                    cnt.pop(s[l])
                l += 1

        return res

File: 1358. Number of Substrings Containing All Three Characters/main.go
package main

func numberOfSubstrings(s string) int {
	res := 0
	l := 0
	cnt := make(map[rune]int)

	for r, char := range s {
		cnt[char]++
		for len(cnt) == 3 {
			res += len(s) - r
			lChar := rune(s[l])
			cnt[lChar]--
			if cnt[lChar] == 0 {
				delete(cnt, lChar)
			}
			l++
		}
	}

	return res
}
